File: tool/sync_shops.py
#!/usr/bin/env python3
"""Google Sheets CSVをアプリ同梱の店舗JSONへ変換する。"""

from __future__ import annotations

import argparse
import csv
import json
import re
import sys
import time
import urllib.parse
import urllib.request
from pathlib import Path


REPO_ROOT = Path(__file__).resolve().parents[1]
DEFAULT_OUTPUT = REPO_ROOT / "assets" / "shops.json"
DEFAULT_CACHE = REPO_ROOT / "tool" / "geocode_cache.json"
EXCLUDED_STATUSES = {"閉店", "閉店予定"}


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("csv_path", type=Path)
    parser.add_argument("--output", type=Path, default=DEFAULT_OUTPUT)
    parser.add_argument("--cache", type=Path, default=DEFAULT_CACHE)
    parser.add_argument("--no-geocode", action="store_true")
    return parser.parse_args()


def load_cache(path: Path) -> dict[str, dict[str, float] | None]:
    if not path.exists():
        return {}
    return json.loads(path.read_text(encoding="utf-8"))


def save_json(path: Path, value: object) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(
        json.dumps(value, ensure_ascii=False, indent=2) + "\n",
        encoding="utf-8",
    )


def geocode(address: str) -> dict[str, float] | None:
    candidates = [address]
    without_parentheses = re.sub(r"[（(].*?[）)]", "", address).strip()
    without_building = without_parentheses.split(" ", 1)[0].strip()
    for candidate in (without_parentheses, without_building):
        if candidate and candidate not in candidates:
            candidates.append(candidate)

    for candidate in candidates:
        params = urllib.parse.urlencode({"q": candidate})
        request = urllib.request.Request(
            f"https://msearch.gsi.go.jp/address-search/AddressSearch?{params}",
            headers={
                "User-Agent": "bento-navi-data-sync/1.0 (github.com/Hammy-the-cat/bento-navi)"
            },
        )
        try:
            with urllib.request.urlopen(request, timeout=30) as response:
                results = json.loads(response.read().decode("utf-8"))
            if results:
                lon, lat = results[0]["geometry"]["coordinates"]
                return {"lat": float(lat), "lon": float(lon)}
        except Exception:
            continue
    return None


def combine_notes(row: dict[str, str]) -> str:
    parts = [row.get("予約・配達メモ", "").strip(), row.get("備考", "").strip()]
    return " ".join(part for part in parts if part)


def main() -> int:
    if hasattr(sys.stdout, "reconfigure"):
        sys.stdout.reconfigure(encoding="utf-8")
    args = parse_args()
    cache = load_cache(args.cache)
    shops: list[dict[str, object]] = []
    unresolved: list[tuple[str, str, str]] = []
    geocode_requests = 0

    with args.csv_path.open(encoding="utf-8-sig", newline="") as csv_file:
        rows = list(csv.DictReader(csv_file))

    for row in rows:
        status = row.get("確認状態", "").strip()
        if not row.get("店舗ID", "").strip() or status in EXCLUDED_STATUSES:
            continue

        address = row.get("住所", "").strip()
        lat_text = row.get("緯度", "").strip()
        lon_text = row.get("経度", "").strip()
        coords: dict[str, float] | None = None
        if lat_text and lon_text:
            coords = {"lat": float(lat_text), "lon": float(lon_text)}
        elif address and "要確認" not in address and not args.no_geocode:
            if address not in cache or cache[address] is None:
                if geocode_requests:
                    time.sleep(0.3)
                print(f"geocode: {row['店舗ID']} {address}", flush=True)
                cache[address] = geocode(address)
                geocode_requests += 1
                save_json(args.cache, cache)
            coords = cache[address]

        if coords is None:
            unresolved.append((row.get("店舗ID", ""), row.get("店舗名", ""), address))
            continue

        shops.append(
            {
                "id": row.get("店舗ID", "").strip(),
                "name": row.get("店舗名", "").strip(),
                "category": row.get("カテゴリ", "").strip(),
                "municipality": row.get("市町村", "").strip(),
                "address": address,
                "lat": coords["lat"],
                "lon": coords["lon"],
                "phone": row.get("電話番号", "").strip(),
                "hours": row.get("営業時間", "").strip(),
                "closedDays": row.get("定休日", "").strip(),
                "notes": combine_notes(row),
                "sourceUrl": row.get("情報源URL", "").strip(),
                "status": status,
                "lastVerified": row.get("最終確認日", "").strip(),
                "coordinateAccuracy": (
                    row.get("座標精度", "").strip()
                    if lat_text and lon_text
                    else "国土地理院住所検索座標（アプリ反映時）"
                ),
            }
        )

    shops.sort(key=lambda shop: str(shop["id"]))
    save_json(args.output, shops)
    save_json(args.cache, cache)

    print(f"written: {len(shops)} shops -> {args.output}")
    print(f"excluded closed: {sum(1 for row in rows if row.get('確認状態', '').strip() in EXCLUDED_STATUSES)}")
    print(f"new geocode requests: {geocode_requests}")
    print(f"unresolved: {len(unresolved)}")
    for store_id, name, address in unresolved:
        print(f"  {store_id}\t{name}\t{address}")
    return 0 if not unresolved else 2


if __name__ == "__main__":
    raise SystemExit(main())

File: tool/sync_shops.mjs
#!/usr/bin/env node

import fs from 'node:fs';
import path from 'node:path';
import process from 'node:process';
import { execFileSync } from 'node:child_process';

const HEADER_ALIASES = new Map([
  ['店舗ID', ['店舗ID', 'id']],
  ['店舗名', ['店舗名', 'name']],
  ['カテゴリ', ['カテゴリ', 'category']],
  ['市町村', ['市町村', 'municipality']],
  ['住所', ['住所', 'address']],
  ['緯度', ['緯度', 'latitude']],
  ['経度', ['経度', 'longitude']],
  ['電話番号', ['電話番号', 'phone']],
  ['営業時間', ['営業時間', 'opening_hours']],
  ['定休日', ['定休日', 'closed_days']],
  ['予約・配達メモ', ['予約・配達メモ', 'delivery_notes']],
  ['情報源URL', ['情報源URL', 'url']],
  ['最終確認日', ['最終確認日', 'checked']],
  ['確認状態', ['確認状態', 'status']],
  ['座標精度', ['座標精度', 'coordinate_accuracy']],
  ['備考', ['備考', 'note']],
]);

const REQUIRED_HEADERS = [...HEADER_ALIASES.keys()];

const EXCLUDED_STATUSES = new Set(['閉店', '閉店予定']);

function parseCsv(text) {
  const rows = [];
  let row = [];
  let cell = '';
  let quoted = false;

  for (let i = 0; i < text.length; i += 1) {
    const char = text[i];
    if (quoted) {
      if (char === '"' && text[i + 1] === '"') {
        cell += '"';
        i += 1;
      } else if (char === '"') {
        quoted = false;
      } else {
        cell += char;
      }
    } else if (char === '"') {
      quoted = true;
    } else if (char === ',') {
      row.push(cell);
      cell = '';
    } else if (char === '\n') {
      row.push(cell.replace(/\r$/, ''));
      rows.push(row);
      row = [];
      cell = '';
    } else {
      cell += char;
    }
  }

  if (cell.length > 0 || row.length > 0) {
    row.push(cell.replace(/\r$/, ''));
    rows.push(row);
  }
  return rows;
}

function value(row, indexes, header) {
  return (row[indexes.get(header)] ?? '').trim();
}

function resolveIndexes(headers) {
  const rawIndexes = new Map(
    headers.map((header, index) => [header.replace(/^\uFEFF/, '').trim(), index]),
  );
  const indexes = new Map();
  for (const [canonical, aliases] of HEADER_ALIASES) {
    const alias = aliases.find((candidate) => rawIndexes.has(candidate));
    if (alias) indexes.set(canonical, rawIndexes.get(alias));
  }
  return indexes;
}

function combineNotes(memo, remarks) {
  return [memo, remarks].filter(Boolean).join(' ');
}

function coordinateAccuracy(sheetValue, previousValue) {
  if (!sheetValue || sheetValue === '未取得') {
    return previousValue || sheetValue || '';
  }
  return sheetValue;
}

function loadExisting(outputPath, baselineRef) {
  if (baselineRef?.startsWith('git:')) {
    return JSON.parse(
      execFileSync('git', ['show', baselineRef.slice(4)], { encoding: 'utf8' }),
    );
  }
  return fs.existsSync(outputPath)
    ? JSON.parse(fs.readFileSync(outputPath, 'utf8'))
    : [];
}

function main() {
  const args = process.argv.slice(2);
  const mergeExisting = args.includes('--merge-existing');
  const positionalArgs = args.filter((arg) => arg !== '--merge-existing');
  const csvPath = positionalArgs[0];
  const outputPath = positionalArgs[1] ?? 'assets/shops.json';
  const baselineRef = positionalArgs[2];
  if (!csvPath) {
    console.error(
      'Usage: node tool/sync_shops.mjs <store-master.csv> [output.json] [git:<ref>:<path>] [--merge-existing]',
    );
    process.exit(1);
  }

  const rows = parseCsv(fs.readFileSync(csvPath, 'utf8'));
  const headers = rows.shift() ?? [];
  const indexes = resolveIndexes(headers);
  const missingHeaders = REQUIRED_HEADERS.filter((header) => !indexes.has(header));
  if (missingHeaders.length > 0) {
    throw new Error(`Missing CSV headers: ${missingHeaders.join(', ')}`);
  }

  const existing = loadExisting(outputPath, baselineRef);
  const existingById = new Map(existing.map((shop) => [shop.id, shop]));
  const shops = [];
  const unresolved = [];

  for (const row of rows) {
    const id = value(row, indexes, '店舗ID');
    const status = value(row, indexes, '確認状態');
    if (!id || EXCLUDED_STATUSES.has(status)) continue;

    const previous = existingById.get(id);
    const latText = value(row, indexes, '緯度');
    const lonText = value(row, indexes, '経度');
    const sheetLat = latText ? Number(latText) : undefined;
    const sheetLon = lonText ? Number(lonText) : undefined;
    const preservePreviousCoordinates =
      previous &&
      Number.isFinite(sheetLat) &&
      Number.isFinite(sheetLon) &&
      Number.isFinite(previous.lat) &&
      Number.isFinite(previous.lon) &&
      Math.abs(sheetLat - previous.lat) < 0.00001 &&
      Math.abs(sheetLon - previous.lon) < 0.00001;
    const lat = preservePreviousCoordinates
      ? previous.lat
      : Number.isFinite(sheetLat)
        ? sheetLat
        : previous?.lat;
    const lon = preservePreviousCoordinates
      ? previous.lon
      : Number.isFinite(sheetLon)
        ? sheetLon
        : previous?.lon;
    if (!Number.isFinite(lat) || !Number.isFinite(lon)) {
      unresolved.push({ id, name: value(row, indexes, '店舗名') });
      continue;
    }

    shops.push({
      id,
      name: value(row, indexes, '店舗名'),
      category: value(row, indexes, 'カテゴリ'),
      municipality: value(row, indexes, '市町村'),
      address: value(row, indexes, '住所'),
      lat,
      lon,
      phone: value(row, indexes, '電話番号'),
      hours: value(row, indexes, '営業時間'),
      closedDays: value(row, indexes, '定休日'),
      notes: combineNotes(
        value(row, indexes, '予約・配達メモ'),
        value(row, indexes, '備考'),
      ),
      sourceUrl: value(row, indexes, '情報源URL'),
      status,
      lastVerified: value(row, indexes, '最終確認日'),
      coordinateAccuracy: coordinateAccuracy(
        value(row, indexes, '座標精度'),
        previous?.coordinateAccuracy,
      ),
    });
  }

  if (mergeExisting) {
    const importedById = new Map(shops.map((shop) => [shop.id, shop]));
    const existingIds = new Set(existing.map((shop) => shop.id));
    const merged = existing.map((shop) => importedById.get(shop.id) ?? shop);
    for (const shop of shops) {
      if (!existingIds.has(shop.id)) merged.push(shop);
    }
    shops.length = 0;
    shops.push(...merged);
  } else {
    shops.sort((a, b) => a.id.localeCompare(b.id, 'ja'));
  }
  fs.mkdirSync(path.dirname(outputPath), { recursive: true });
  fs.writeFileSync(outputPath, `${JSON.stringify(shops, null, 2)}\n`, 'utf8');

  console.log(
    JSON.stringify(
      {
        written: shops.length,
        output: outputPath,
        unresolved,
      },
      null,
      2,
    ),
  );
}

main();
